Encryption support for Cilium

jrfastab opened the pull request:
Initial Cilium support for IPSec, only tunnel support in this series direct routing will be added shortly.

Todo for this PR:
- ~~fix pinging between cilium host ips .e.g 156.10.0.1 -> 157.10.0.1~~
- ~~pull in keys from secrets file (pushing patches now)~~
- ~~rebase~~
- ~~fixup hound comments~~
- ~~forgot to update unit tests with new APIs currently failing make tests.~~
- ~~move enable-ipsec flag earlier in stack and only add routes/rules when enabled~~

Series behind this: (really just to stop this series from growing to unreviewable chunk [edit: And also so I can stop rebasing behind everyone.])
- add some unit tests
- enable esn
- IPv6 support
- Improved error path messages
- direct routing hooks
- Drop tunnel mode ipsec in place of transport should almost work as is but tunnel mode was a bit easier to dbg IMO but can fix with bpf tracing.
- Add monitor support so we can see whats happening.
- On 4.14+ kernels push security label into ip options for direct routing and vxlan modes
- Documentation
- per endpoint strict mode to drop unencrypted traffic.
- push iptables pkg into linux datapath
- build list of marks that can be imported multiple places without pulling in all of datapath/proxy

There are some more fun series to work on but above is the min viable for deployment with out "testing" flag IMO.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6693)
<!-- Reviewable:end -->

jrfastab committed:
cilium: ipsec, report flags with endpoint maps

Reporting flags with endpoint maps helps debugging by allocwing us to
ensure the datapath and control path are in sync. Useful when ipsec
becomes enabled and host/endpoint flags determine IPSec flows.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, extend datapath route pkg to support more specifiers

In order to support IPSec we use specific routes inserted into a
table designated for IPSec. To further isolate these entries we
also constrain the routes using the proto (proto=50 for ESP) field
and insert as a "local" route to ensure traffic is processed by the
IPSec in kernel code on the correct device. Details on IPSec
implementation in documentation and subsequent patches.

To facilitate this add support in route pkg to program these from
the API.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add support for inserting rules via golang API

Currently, existing route rules are inserted via init script. Because we are
going to start adding more rules for IPSec this patch adds a proper route API
that we can call directly.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add go API to configure xfrm (IPSec)

This adds a new datapath directory ipsec to configure ipsec for Cilium.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, insert 'ip rules' for ipsec

IPSec will be isolated from the "normal" routing tables via a set of
marks to direct traffic to an IPSec table. When an endpoint has IPSec
enabled the datapath will first set the mark bit to endicate
MARK_ENCRYPTED or MARK_DECRYPTED to depending on expected direction.
Once the encryption or decryption has occured the mark fields will
be cleared and normal processing will happen.

In this patch add the mark based 'ip rules' to segregate the IPSec
traffic.

Also decided on adding default const to ./pkg/datapath/linux/datapath.go
becaause (a) as defaults they only make sense on linux datapath it
seems and (b) originally I pushed them in via ./pkg/option/ and
./pkg/defaults but this got a bit ugly IMO to translate through to
lower datapath after latest changes to separate fake/linux dp. Sure
we can push them through NewDaemon but do Daemon creators really want
this control? Anyways its easy to go back to pushing at Daemon
create time if anyone cares.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add ip route needed to decrypt/encrypt traffic

IPSec datapath pushes packets into IP stack for encryption and decryption.
This adds the routing rules used to enforce this. These are inserted once
per node so overhead should not be concerning.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, qualify iptables rules to avoid colliding with ipsec

Cilium uses a handful of IPTable rules to mark local packets to/from the proxy
along with a few other cases. We need to ensure that these rules do not
interfere with packets that need to be encrypted or decrypted. This patch adds
qualifiers to the iptables rules added to ensure the encrypt/decrypt mark bits
are not set before taking any actions.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add BPF datapath encryption direction

This enables encryption on TX path for traffic that is sent over a
vxlan tunnel.

Currently, we use tunneled IPSec but the datapath should work regardless.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, bpf fix mark comments and func names to reflect current usage

The current skb->mark comment is out of date and no longer reflects the actual
usage of the 32bit mark field. Fix this and generalize get_identity_via_proxy
to simply get_identity it will be used in IPSec case now as well.

The cluster ID was removed here, 661062fec8a1 ("bpf: Remove cluster identity").

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, add option and #define for ENABLE_IPSEC enablement

Add a command line option to enable inter-cilium-node IPSec encryption.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, bpf add receive side ipsec

If we did nothing here cilium_vxlan will pop the vxlan header then
the outer header will route to cilium_host (assuming tunnel mode)
this route would hit the ipsec decrypt path so cilium host would
get decrypted packet. These would process using normal host logic
and be locally redirected to lxc endpoint via BPF program. However,
when mixed with non-ipsec traffic some flows were getting
encrypted/decrypted unexpectedly.

So instead of hopping through multiple BPF programs this changes the
flow as follows. Set the mark decrypt bit and pass to stack. The
routing stack decrypts the packet and routes it back to cilium
vxlan after decryption is complete. After this cilium_vxlan "sees"
the decrypted bit is set and does redirect to lxc. We keep the
ipsec traffic localized this way. IMO its easier to trace the
flow.

We had to pull in a portion of if_packet to get magic numbers
for PACKET_* values. Only pulled in required defines for now.

Finally, in the next set of patches we drop the tunnel mode flag
and this becomes even more relavent to avoid (re)routing.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: ipsec, import keys via certs file

Initially, we will support a pre-shared key model where all endpoints
have a pre-shared keys loaded from a file. The file layout is as
follows with a key per line.

authname authkey encname enckey [scope]

This allows keys to be scoped so that it is possible to have a key
per node if desired. If scope is omitted the key is considered a
default key used if a more specific key is not found. My three node
testing file reads as follows,

hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF ubuntu-kvm1
hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF ubuntu-kvm2
hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF

The file is loaded at init time so any changes will have to be consumed
by a restart. In the future we can add watchers for the file and allow
using etcd secrets (files).

Also note the logic is built so that other methods of key exchanges
can be added later if folks want a different model.

Looking forward we can more specific scope types to do per identity
keys if needed.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on Documentation/cmdref/cilium-bugtool.md:
````diff
@@ -6,10 +6,11 @@ Collects agent & system information useful for bug reporting
 
 ```
-cilium-bugtool [OPTIONS] [flags]
+cilium-bugtool [OPTIONS]
````
Not sure what's going on here, but when I run `make -C Documentation cmdref` on top of this series, this gets reversed again.

You might need to do a `make clean` and `make` again to flush some cmdref cache (and drop this particular patch).

jrfastab replied:
yep dropped seems fixed by clean/make cycle.

jrfastab commented on pkg/datapath/linux/node.go:
```diff
@@ -39,6 +39,10 @@ type linuxNodeHandler struct {
 }
 
+var (
+	RTN_LOCAL = 0x2 // Define RTN_LOCAL to avoid another import
```
I'm ignoring this its not searchable in includes and into C BPF files if its lowercase.

joestringer commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -50,6 +53,8 @@ func (r *Route) getNetlinkRoute() netlink.Route {
 
 	if r.Scope != 0 {
 		rt.Scope = r.Scope
+	} else if r.Scope == 0 && r.Type == 2 {
```
What are these magic numbers? `unix.RT_SCOPE_UNIVERSE` and `unix.RTN_LOCAL`?

jrfastab replied:
yep.

joestringer commented on pkg/datapath/linux/datapath.go:
```diff
@@ -18,6 +18,29 @@ import (
+	// RouteMarkMask is the default mask to use when reading RouteMark
+	RouteMarkMask = 0x0F00
+
+	// IPSecProtocolID IP protocol ID for IPSec
```
Might be nice to leave a comment stating this is in reference to RFC4303.

jrfastab replied:
sure.

joestringer commented on pkg/datapath/linux/datapath.go:
```diff
@@ -18,6 +18,29 @@ import (
+	RouteMarkEncrypt = 0x0E00
+
+	// RouteMarkMask is the default mask to use when reading RouteMark
+	RouteMarkMask = 0x0F00
```
I assume that the above constants were chosen by using the next available marks after the ones defined in `pkg/proxy/mark.go`? (We should probably fold these definitions together so it's easier to tell which marks we're using for what).

jrfastab replied:
Would be nice but the naive implementation being to simply add them to the mark.go results in another import in the datapath side. And having the datapath import proxy is ugly IMO vice versa is just as ugly. How about putting this in a TODO. I agree with the idea however.

joestringer replied:
Arguably the proxy setup details is part of the datapath as well, but yeah that requires a bit more work to refactor. Could consider adding a `pkg/datapath/constants` as an intermediate if it's useful. But yeah let's punt this for now.

jrfastab replied:
My opinion is we have datapath/constants and move iptables and proxy under datapath but lets do that as followup work thanks.

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	policy.Mark = &netlink.XfrmMark{
+		Value: 0x0D00,
+		Mask:  0x0F00,
```
Hardcoded magic numbers

jrfastab replied:
sure resolved by pushing linux datapath constants earlier in series and using them here.

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	policy.Mark = &netlink.XfrmMark{
+		Value: 0x0E00,
+		Mask:  0x0F00,
```
Hardcoded magic numbers

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+	}
+	policy.Dst = &net.IPNet{
+		IP:   localCilium,
+		Mask: net.IPv4Mask(255, 255, 0, 0),
```
I assume these are traffic matching conditions? Should we make the prefix configurable from a single point?

Presumably there are places we can get at least some of this information from - for instance, the local IPs are typically within a /24 but we know what the CIDR range is because we need to allocate IPs from within that range.

For the remote peer it might be a bit different, I forget how exactly we apply that these days but maybe just having a global configuration for that as well with a reasonable default prefix would be fine.

jrfastab replied:
Its available the only thought is do it in this series or on top.

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+	Crypt *netlink.XfrmStateAlgo
+}
+
+var ipSecKeysGlobal = make(map[string]*ipSecKey)
```
it would appear that concurrent access of this is considered safe because it can only be configured via `LoadIPSecKeysFile()` and that is protected by the global compilation lock; Maybe that's worth documenting somewhere?

jrfastab replied:
right, only writers are at daemon init time so readers are free to read unlocked. I added a comment.

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+ *
+ * For all traffic that matches a policy, the policy tuple used is
+ * (sip/mask, dip/mask, dev) with an optional mark field used in the Cilium implementation
+ * to ensure only expected traffic is encrypted, the state hashtable is searched for
```
```suggestion
 * to ensure only expected traffic is encryptede The state hashtable is searched for
```

jrfastab replied:
fixed

joestringer replied:
heh, even my correction had a typo -_-"

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+ * state space. Basic idea would be to reference a state using any key generated
+ * from BPF program allowing for a single state per security ctx.
+ */
+func UpsertIPSecEndpoint(endpoint net.IPNet, ciliumIP net.IP) error {
```
what's the endpoint vs the ciliumIP here? `endpoint` appears unused.

joestringer replied:
To answer my own question: The `endpoint` should be an actual endpoint, ie container IP while the `ciliumIP` is presumably the local host IP for the specified endpoint. Hence the big todo below about duplication, because if there is more than one endpoint on a peer node then it is guaranteed to print a log (cf. `OnIPIdentityCacheChange()`)

jrfastab replied:
right. The question is do we want to create a map (and associated memory) to track this when the kernel already "knows" the state. What log msg are you mentioning above? Hopefully I'm not creating a bunch of ugly log messages. If that is true maybe we can at minimum silence those.

joestringer replied:
I suspect that every time you learn about a new endpoint on a peer node, it will attempt this so for endpoitnts 2+ on each peer node it would generate this log.

jrfastab replied:
The map will need to be the sized to align with number of nodes.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -417,6 +417,11 @@ func init() {
 	option.BindEnv(option.EnvoyLog)
 
+	flags.Bool(option.EnableIPSecName, defaults.EnableIPSec, "Enable IPSec support")
+	viper.BindEnv(option.EnableIPSecName, option.EnableIPSecNameEnv)
```
```suggestion
	option.BindEnv(option.EnableIPSecName)
```

joestringer commented on pkg/option/config.go:
```diff
@@ -373,6 +373,10 @@ const (
+
+	// EnableIPSecName is the name of the option to enable IPSec
+	EnableIPSecName    = "enable-ipsec"
+	EnableIPSecNameEnv = "CILIUM_ENABLE_IPSEC"
```
```suggestion
```

(`option.BindEnv()` will take care of this for you.)

jrfastab replied:
great

joestringer commented on daemon/daemon_main.go:
```diff
@@ -417,6 +417,11 @@ func init() {
+	flags.Bool(option.EnableIPSecName, defaults.EnableIPSec, "Enable IPSec support")
+	viper.BindEnv(option.EnableIPSecName, option.EnableIPSecNameEnv)
+
+	flags.StringVar(&option.Config.IPSecKeyFile, "ipsec-key-file", "", "Path to IPSec key file")
```
```suggestion
	flags.StringVar(&option.Config.IPSecKeyFile, "ipsec-key-file", "", "Path to IPSec key file")
	option.BindEnv(option.IPSecKeyFile)
```

joestringer commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,300 @@
+
+		s := strings.Split(scanner.Text(), " ")
+		if len(s) < 4 {
+			return fmt.Errorf("missing IPSec keys or invalid format")
```
You might consider printing the line with the invalid format here, for easier debug.

joestringer commented on pkg/identity/identity.go:
```diff
@@ -52,6 +52,7 @@ type IPIdentityPair struct {
 	IP       net.IP          `json:"IP"`
 	Mask     net.IPMask      `json:"Mask"`
 	HostIP   net.IP          `json:"HostIP"`
+	CiliumIP net.IP          `json:"CiliumIP"`
```
What's the difference between `HostIP` and `CiliumIP`?

The description in this commit describes it as the `cilium host IP` which just adds to the confusion.

While the wire format for `HostIP` probably can't be changed without introducing some form of interoperation problem (eg, connectivity outage during upgrade), perhaps we can use a more precise name, eg `ExternalHostIP` / `HostIPExternal` for the new field? And add some comments somewhere here to explain the rationale.

jrfastab replied:
HostIP is the externalHostIP the one you might see to ping the node, the CiliumIP is the ip assigned to cilium_host.

joestringer replied:
Ah, okay, I had it backwards. I guess that's the origin of the name `ciliumIP`, as it's associated with `cilium_host`.

jrfastab replied:
Right I tried to make the comment easier to parse maybe it helped.

joestringer commented on pkg/ipcache/kvstore.go:
```diff
@@ -165,13 +165,14 @@ func (r *kvReferenceCounter) release(key string) (err error) {
 		ID:       ID,
 		Metadata: metadata,
 		HostIP:   hostIP,
+		CiliumIP: ciliumIP,
```
Not really a comment for this PR, but it looks like this `IPIdentityPair` is getting a lot more involved than just a pair of ip and its corresponding identity..

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -413,6 +417,19 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *node.Node) error {
+	if n.nodeConfig.EnableIPSec {
+		new4Net := &net.IPNet{IP: newIP4, Mask: newNode.IPv4AllocCIDR.Mask}
+		new6Net := &net.IPNet{IP: newIP6, Mask: newNode.IPv6AllocCIDR.Mask}
+		n.replaceHostRules()
```
One of the big changes recently in this area was to shift away from replacing all rules every time a node is updated. I don't know how big this table is, but it seems like this change has the potential to unwind some of those gains.

jrfastab replied:
maybe obvious we don't need to replaceHostRules on every addition at least :/ thinking...

joestringer replied:
On the plus side this is still gated by `EnableIPSec`, so this could be addressed in a future PR.

jrfastab replied:
This code above I believe should only be called "NodeConfigurationChanged is called when the LocalNodeConfiguration has changed" which I think should be the correct context for above code.

But in general I agree with the comment. Currently the code to add the routes to the kernel are listening on the ipcache so any entry in there results in touching the route update kernel side.

joestringer replied:
OK, I think we're on the same page. My comment is perhaps a bit more about the implementation underlying `replaceHostRules` in that it's a "replace everything" rather than making a specific add/delete, and the replace appears to look through potentially many rules.

jrfastab replied:
Sure although at the moment there is only two rules so doing a replace on two rules shouldn't hurt much. I expect LocalNodeConfiguration changes are relatively infrequent.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -495,6 +512,89 @@ func (n *linuxNodeHandler) updateOrRemoveClusterRoute(addressing datapath.NodeAd
+	}
+}
+
+// replaceNodeIPSecoOutRoute verifies whether the node route is properly covered by a
```
(nit) The comments for this function and its sister function suggest some kind of "verifying", but there's no return condition that suggests this. Seems like it just does a replace, like the function name says.

joestringer replied:
```suggestion
// replaceNodeIPSecOutRoute verifies whether the node route is properly covered by a
```

joestringer commented on pkg/datapath/linux/datapath.go:
```diff
@@ -18,6 +18,29 @@ import (
+
+	// RouteMarkEncrypt is the default route mark to use to indicate datapath
+	// needs to encrypt a packet.
+	RouteMarkEncrypt = 0x0E00
```
I guess that this PR doesn't address encrypting proxy yet, but it would be good to double-check stuff like `MagicMarkProxyMask`; with the value `0xE00` it sounds like we may end up considering somewhere that the new encrypted/decrypted traffic is actually proxy traffic. Might need to update some of the proxy checks.

jrfastab replied:
Right not addressing proxy yet on egress. I was addressing proxy in a follow up series.

joestringer commented on bpf/lib/common.h:
```diff
@@ -286,29 +286,41 @@ enum {
- *    an ingress or egress proxy, or a local process.
- *  - the cluster id
+ * The upper 16 bits plus lower 8 bits (e.g. mask 0XFFFF00FF) contain the
+ * packets security idetity. The lower/upper halves are swapped to recover
```
```suggestion
 * packets security identity. The lower/upper halves are swapped to recover
```

joestringer commented on bpf/lib/common.h:
```diff
@@ -286,29 +286,41 @@ enum {
+ * The 4 bits at 0XF00 provide
+ *  - the magic marker values which indicate whether the packet is coming from
+ *    an ingress or egress proxy, a local process and its current encryption
+ *    status.
```
This comment exactly reverses the description of the bits in terms of `upper` vs. `lower`... I think it's meant to describe the mark when the packet is received into Cilium. Part of the confusion looks like it's that `get_identity()` actually unscrambles the identity so that it returns it as a contiguous 24 bits (ie bits may be set within the range 0x00FFFFFF), however the actual mark itself still has this weird reverse structure that was described in the previous comment. (Either that or we actually have differing understandings of upper/lower..)

At least, that's my reading from the following function `get_identity()` which retrieves the identity from the mark:

https://github.com/cilium/cilium/blob/867881a7c2efcb813ce220c7b55d2f9b86cba31e/bpf/lib/common.h#L309

joestringer commented on pkg/datapath/ipcache/listener.go:
```diff
@@ -115,6 +116,16 @@ func (l *BPFListener) OnIPIdentityCacheChange(modType ipcache.CacheModification,
 		}
+
+		if value.SecurityIdentity > identity.ReservedIdentityUnmanaged.Uint32() {
+			err = ipsec.UpsertIPSecEndpoint(cidr, ciliumIP)
```
The inverse of this operation seems to never be applied when the ipcache entry is deleted, I assume this is because other endpoints may still exist on the remote node so we still need the IPSec configuration applied for communication with other endpoints on that node.

This might suggest that the `OnIPIdentityCacheChange` interface isn't quite the right place for this long-term. That said, what are the consequences of this? I guess that if someone configures IPsec then unconfigures it, then we might not unconfigure it on the kernel side so something might break? Or would the xfrm state be cleared out on daemon startup?

jrfastab replied:
As for location, early on in this series I had per endpoint keys so every endpoint could possibly use its own key. This was a bit much for a first iteration even if we eventually get there on a per service basis for example. So I've left it hooked into ipcache so that we get these events. Anyways I'm not sure if we have a "endpoint subnet" has come online event anywhere.

On daemon startup we can flush the state/policy. I'll add this code now. This seems to be inline with other configuration we setup with Cilium. In general we don't tear it down on cilium exit but on daemon bring up.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -635,8 +635,15 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
+		/* If not redirected noteably due to IPSEC then pass up to stack
+		 * for further processing.
+		 */
+		goto pass_to_stack;
```
What happens if `ret` is an error? Shouldn't the IPSEC case be clearly differentiated from errors or successful redirects?

jrfastab replied:
hmm OK we should catch DROP_WRITE error case as well. Then I think we do not need to do anything to differentiate between ipsec and non-ipsec. So I think the error here is not properly handling the error cases and dropping pkt.

jrfastab replied:
```
                /* Redirected by encap code path */
                if (ret == TC_ACT_REDIRECT)
                        return ret;
                /* If not redirected noteably due to IPSEC then pass up to stack
                 * for further processing.
                 */
                else if (!ret)
                        goto pass_to_stack;
                /* Otherwise this is an error and packet is being dropped. */
                else
```

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -651,11 +658,18 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
+		 */
+		ret = encap_and_redirect(skb, &key, SECLABEL, monitor, false);
+		if (ret == DROP_NO_TUNNEL_ENDPOINT)
+			goto to_host;
```
AFAICT the previous version of this fell through to `pass_to_stack` (via the goto immediately following this `#ifdef` macro section), not `to_host`. Again, I think we need to call out the ipsec case more clearly here because `ret` could also be a drop-type error, in which case it seems like we should drop and not just pass it to the stack.

joestringer replied:
(for context, I'm looking at the non-IPsec `encap_and_redirect_with_nodeid()` case)

jrfastab replied:
```
#ifdef ENCAP_IFINDEX
...
        } else {
...
                ret = encap_and_redirect(skb, &key, SECLABEL, monitor);
    
                /* Fall through if remote prefix was not found
                 * (DROP_NO_TUNNEL_ENDPOINT) */
                if (ret != DROP_NO_TUNNEL_ENDPOINT)
                        return ret;
        }
#endif 
        goto pass_to_stack;

to_host:
```
so seems to_host -> pass_to_stack should fix it. Good catch.

```
                ret = encap_and_redirect(skb, &key, SECLABEL, monitor, false);
                if (ret == DROP_NO_TUNNEL_ENDPOINT)
                        goto pass_to_stack;
                else if (!ret)
                        goto pass_to_stack;
                else
                        return ret;
```

jrfastab replied:
how about above ^^^^ this introduces !ret as the IPSEC case basically. Could give it a naem I guess, IPSEC_TUNNEL_ENDPOINT so that its a bit cleaner,
```
else if (ret == IPSEC_TUNNEL_ENDPOINT)
```

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -512,6 +512,21 @@ int from_netdev(struct __sk_buff *skb)
+			return encap_and_redirect_with_nodeid(skb, tunnel_endpoint, seclabel, TRACE_PAYLOAD_LEN);
+		}
+	}
+#endif
```
Can we get a `send_trace_notify()` here (cf. later in the function)? Looks like we have no other way to `cilium monitor` this traffic that gets redirected to another node from local applications.

jrfastab replied:
figure this can come on its own patch series it needs some extra defines etc. Can add it to this series if you want. I would prefer a new series though.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -351,7 +351,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 		key.family = ENDPOINT_KEY_IPV6;
 
-		ret = encap_and_redirect(skb, &key, SECLABEL, monitor);
+		ret = encap_and_redirect(skb, &key, SECLABEL, monitor, true);
```
The corresponding IPv4 case sets the `host` bool to `false`, is this supposed to be `false`?

joestringer commented on bpf/lib/encap.h:
```diff
@@ -46,16 +65,31 @@ encap_and_redirect_with_nodeid(struct __sk_buff *skb, __u32 tunnel_endpoint,
 static inline int __inline__
 encap_and_redirect(struct __sk_buff *skb, struct endpoint_key *k,
-		   __u32 seclabel, __u32 monitor)
+		   __u32 seclabel, __u32 monitor, bool host)
```
nit: `from_host` would be a bit more verbose but also a bit more descriptive.

jrfastab replied:
sure.

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -95,18 +95,48 @@ static inline int handle_ipv4(struct __sk_buff *skb)
+				goto not_esp;
+			skb->mark = MARK_MAGIC_DECRYPT;
+			set_identity(skb, key.tunnel_id);
+			skb_change_type(skb, 0);
```
what's `0` denote here? `PACKET_HOST`?

jrfastab replied:
yep packet_host seems I dropped my comment here after merge conflict resolution. This is a bit tricky because we are telling the host its a packet for the host so the IP stack doesn't drop it on us.

joestringer commented on bpf/lib/encap.h:
```diff
@@ -46,16 +65,31 @@ encap_and_redirect_with_nodeid(struct __sk_buff *skb, __u32 tunnel_endpoint,
 static inline int __inline__
 encap_and_redirect(struct __sk_buff *skb, struct endpoint_key *k,
-		   __u32 seclabel, __u32 monitor)
+		   __u32 seclabel, __u32 monitor, bool host)
```
Any chance we could get a comment above this function describing the expected outputs? It might be useful to have a reference here which states what the return values are and what is expected of the caller.

jrfastab replied:
sure.

joestringer commented on bpf/lib/encap.h:
```diff
@@ -22,6 +22,25 @@
+	 */
+	skb->mark = MARK_MAGIC_ENCRYPT;
+	set_identity(skb, seclabel);
+	skb->cb[4] = tunnel_endpoint;
```
:see_no_evil:

jrfastab replied:
open for better ideas but this works.

joestringer reviewed (changes requested):
I appreciate that you put the PR out at this point rather than later on when it grows even bigger ;-) This was actually really tidy to read commit-by-commit, thanks for putting in the effort there.

I have a mix of cosmetic and functional feedback below, hopefully it is helpful. On the datapath side I noticed a few discrepancies between the existing implementation and the new changes, not sure whether they're intentional. After glancing again at the cover letter I see that once or twice I mentioned stuff like monitor output which you've actually highlighted as something that can come in a future patch.

The biggest caveat I think I see here that I didn't see mentioned in the PR description is that if traffic arrives unencrypted, then we will process that too (no option to force encryption). That's fine to add later for sure, but thought it's worth calling out as it would have implications for anyone wanting to actually try out this feature.

joestringer reviewed (approved):
Changes looks good, but to reduce confusion I think it'd be better to name the CiliumIP in the API (ipcache kvstore entries json formatting) and usage throughout the codebase with internalIP. This is how the node package refers to it, so at least it would be consistent.

aanm commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,260 @@
+	Crypt *netlink.XfrmStateAlgo
+}
+
+var ipSecKeysGlobal = &ipSecKeys{
```
Where is this super secure global key used? Can you add a comment why this is needed?

jrfastab replied:
Its read in from a file. I can add a comment the reason is simply we need keys for IPSec and have to pull them from somewhere. In iteration one here we read them from a file specified on the CLI. Follow up patch can make this a mounted etcd secret file.

jrfastab replied:
we could avoid saving them in memory and read them on-demand if that seems better. For now the file is not encrypted either so it doesn't seem to matter either way.

jrfastab replied:
FWIW the keys are also stored in plaintext in-kernel at the moment so any user with CAP_NET_ADMIN can read them out easily as well so doing a bunch of work here doesn't help much until the kernel side is also fixed to use a KEYRING but that will require a kernel patch.

aanm commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,260 @@
+	ipSecJoinState(state, ipSecKeysGlobal)
+	state.Src = localIP
+	state.Dst = remoteIP
+	err := netlink.XfrmStateAdd(state)
```
small nit you can return the error without assigning to a new variable

jrfastab replied:
yeah I had some old debug code there to log the error but now just wasted line.

aanm commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,260 @@
+
+func ipSecReplacePolicy(endpoint, remoteCilium, localCilium net.IP) {
+	/* Errors are ignored at the moment due to expecting EEXIST errors */
+	ipSecReplacePolicyIn(endpoint, remoteCilium, localCilium)
```
Can't we know if the returned errors is EEXIST?

jrfastab replied:
Sure the comment is a bit dumb. The problem is what to do if it fails. I think throwing a big warning is most likely the best for now.

jrfastab replied:
OK catching error now and passing up stack to deal with.

aanm commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -290,7 +298,9 @@ func InstallRules(ifName string) error {
-			"-o", ifName,
+			"-m", "mark", "!", "--mark", matchFromIPSecDecrypt, // Don't match ipsec traffic
+			"-m", "mark", "!", "--mark", matchFromIPSecEncrypt, // Don't match ipsec traffic
+			"-o", defaults.HostDevice,
```
this needs to be `ifName`

jrfastab replied:
nice catch missed that on rebase. Will fix thanks.

aanm commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -303,7 +313,9 @@ func InstallRules(ifName string) error {
-			"-o", ifName,
+			"-m", "mark", "!", "--mark", matchFromIPSecDecrypt, // Don't match ipsec traffic
+			"-m", "mark", "!", "--mark", matchFromIPSecEncrypt, // Don't match ipsec traffic
+			"-o", defaults.HostDevice,
```
this needs to be `ifName`

jrfastab replied:
yep thanks.

aanm commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -258,3 +259,45 @@ func DeleteIPSecEndpoint(endpointIP net.IP) error {
+			return fmt.Errorf("missing IPSec keys or invalid format")
+		}
+
+		authname, authkey := s[0], []byte(s[1])
```
helps if you add a comment with the file structure this scanner is reading

jrfastab replied:
The layout is in the routine comment. But can also put it closer to the scanner if that helps.

brb commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -288,3 +302,41 @@ func Delete(route Route) error {
 }
+
+func lookupRule(fwmark int, table int) (bool, error) {
+	rules, err := netlink.RuleList(0)
```
Nit: s/0/unix.AF_INET/

jrfastab replied:
hmm its actually AF_UNSPEC == 0 here.

jrfastab replied:
I'm not sure should we try to avoid importing entire pkgs like unix here just to get a single enum? Maybe I'm thinking to hard about it. Here '0' is fairly clear IMO.

brb replied:
Agreed, just maybe add a comment next to it saying `0 =  AF_...` - would help for a reader.

brb commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,303 @@
+	state := netlink.XfrmState{
+		Mode:  netlink.XFRM_MODE_TUNNEL,
+		Proto: netlink.XFRM_PROTO_ESP,
+		ESN:   false,
```
You could enable [ESN](https://tools.ietf.org/html/rfc4304) which allows us to avoid key rotation which is required to mitigate replay attacks.

jrfastab replied:
agree adding it to the todo list.

brb commented on pkg/datapath/ipsec/ipsec_linux.go:
```diff
@@ -0,0 +1,303 @@
+}
+
+/* UpsertIPSecEndpoint updates the IPSec context for a new endpoint inserted in
+ * the ipcache. Currently we support a global crpyt/auth keyset that will encrypt
```
Nit: s/crpyt/crypt/

jrfastab replied:
thanks.

jrfastab committed:
cilium: ipsec, bind nodeUpdate to ipsec upsert to install ipsec state
    
This hooks into nodeUpdate events using it to add/remote IPSec policies
and states when new endpoints come online/offline. For node based
ipsec this is sufficient, if we want finer grained IPSe policies.
Possibly per identity we will need to hook into ipache most likely.
    
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: Add IPv4GW to node discovery

This adds the IPv4 gateway address to the localNode so that on node
events we can use the gateway address to populate the IPSec context.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

ianvernon reviewed (approved):
Approving only for docs.
